metadata-ingestion/src/datahub/ingestion/source/powerbi.py

```diff
@@ -0,0 +1,1382 @@
+#########################################################
+#
+# Meta Data Ingestion From the Power BI Source
+#
+#########################################################
+
+import logging
+from dataclasses import dataclass
+from dataclasses import field as dataclass_field
+from enum import Enum
+from time import sleep
+from typing import Any, Dict, Iterable, List, Optional, Set, Tuple
+from xmlrpc.client import Boolean
+
+import msal
+import requests
+from orderedset import OrderedSet
+
+import datahub.emitter.mce_builder as builder
+from datahub.configuration import ConfigModel
+from datahub.configuration.common import AllowDenyPattern, ConfigurationError
+from datahub.emitter.mcp import MetadataChangeProposalWrapper
+from datahub.ingestion.api.common import PipelineContext
+from datahub.ingestion.api.source import Source, SourceReport
+from datahub.ingestion.api.workunit import MetadataWorkUnit
+from datahub.metadata.com.linkedin.pegasus2avro.common import (
+    ChangeAuditStamps,
+    DataPlatformInstance,
+)
+from datahub.metadata.schema_classes import (
+    BrowsePathsClass,
+    ChangeTypeClass,
+    ChartInfoClass,
+    ChartKeyClass,
+    CorpUserInfoClass,
+    CorpUserKeyClass,
+    DashboardInfoClass,
+    DashboardKeyClass,
+    DataPlatformInfoClass,
+    DatasetKeyClass,
+    DatasetPropertiesClass,
+    OwnerClass,
+    OwnershipClass,
+    OwnershipTypeClass,
+    StatusClass,
+)
+
+# Logger instance
+LOGGER = logging.getLogger(__name__)
+
+
+class Constant:
+    """
+    keys used in powerbi plugin
+    """
+
+    PBIAccessToken = "PBIAccessToken"
+    DASHBOARD_LIST = "DASHBOARD_LIST"
+    TILE_LIST = "TILE_LIST"
+    DATASET_GET = "DATASET_GET"
+    REPORT_GET = "REPORT_GET"
+    DATASOURCE_GET = "DATASOURCE_GET"
+    TILE_GET = "TILE_GET"
+    ENTITY_USER_LIST = "ENTITY_USER_LIST"
+    SCAN_CREATE = "SCAN_CREATE"
+    SCAN_GET = "SCAN_GET"
+    SCAN_RESULT_GET = "SCAN_RESULT_GET"
+    Authorization = "Authorization"
+    WorkspaceId = "WorkspaceId"
+    DashboardId = "DashboardId"
+    DatasetId = "DatasetId"
+    ReportId = "ReportId"
+    SCAN_ID = "ScanId"
+    Dataset_URN = "DatasetURN"
+    CHART_URN = "ChartURN"
+    CHART = "chart"
+    CORP_USER = "corpuser"
+    CORP_USER_INFO = "corpUserInfo"
+    CORP_USER_KEY = "corpUserKey"
+    CHART_INFO = "chartInfo"
+    STATUS = "status"
+    CHART_ID = "powerbi.linkedin.com/charts/{}"
+    CHART_KEY = "chartKey"
+    DASHBOARD_ID = "powerbi.linkedin.com/dashboards/{}"
+    DASHBOARD = "dashboard"
+    DASHBOARD_KEY = "dashboardKey"
+    OWNERSHIP = "ownership"
+    BROWSERPATH = "browsePaths"
+    DASHBOARD_INFO = "dashboardInfo"
+    DATAPLATFORM_INSTANCE = "dataPlatformInstance"
+    DATASET = "dataset"
+    DATASET_ID = "powerbi.linkedin.com/datasets/{}"
+    DATASET_KEY = "datasetKey"
+    DATASET_PROPERTIES = "datasetProperties"
+    VALUE = "value"
+    ENTITY = "ENTITY"
+    ID = "ID"
+
+
+class PowerBiAPIConfig(ConfigModel):
```

Comment:
extend from `EnvBasedSourceConfigBase` in `datahub.configuration.source_common`. 
That will give you the `env` field with a validator.